solvebio/global_search.py

```diff
@@ -0,0 +1,159 @@
+
+import logging
+logger = logging.getLogger('solvebio')
+
+
+class GlobalSearch(Query):
+    """
+    GlobalSearch acts as a request wrapper that generates a request from Filter objects,
+    and can iterate through streaming result sets.
+    """
+
+    def __init__(
+            self,
+            query=None,
+            filters=None,
+            entities=None,
+            ordering=None,
+            limit=float('inf'),
+            page_size=QueryBase.DEFAULT_PAGE_SIZE,
+            result_class=dict,
+            debug=False,
+            **kwargs):
+        """
+        Creates a new Query object.
+
+        :Parameters:
+          - `query` (optional): An optional query string (advanced search).
+          - `filters` (optional): Filter or List of filter objects.
+          - `entities` (optional): List of entity tuples to filter on (entity type, entity).
+          - `ordering` (optional): List of fields to order the results by.
+          - `limit` (optional): Maximum number of query results to return.
+          - `page_size` (optional): Number of results to fetch per query page.
+          - `result_class` (optional): Class of object returned by query.
+          - `debug` (optional): Sends debug information to the API.
+        """
+        self._data_url = '/v2/search'
+        self._query = query
+        self._entities = entities
+        self._ordering = ordering
+        self._result_class = result_class
+        self._debug = debug
+        self._error = None
+        self._is_join = False
+
+        if filters:
+            if isinstance(filters, Filter):
+                filters = [filters]
+        else:
+            filters = []
+        self._filters = filters
+
+        # init response and cursor
+        self._response = None
+        # Limit defines the total number of results that will be returned
+        # from a query involving 1 or more pagination requests.
+        self._limit = limit
+        # Page size/offset are the low level API limit and offset params.
+        self._page_size = int(page_size)
+        # Page offset can only be set by execute(). It is always set to the
+        # current absolute offset contained in the buffer.
+        self._page_offset = None
+        # slice is set when the Query is being sliced.
+        # In this case, __iter__() and next() will not
+        # reset the page_offset to 0 before iterating.
+        self._slice = None
+
+        # parameter error checking
+        if self._limit < 0:
+            raise Exception('\'limit\' parameter must be >= 0')
+
+        if not 0 < self._page_size <= self.MAX_PAGE_SIZE:
+            raise Exception('\'page_size\' parameter must be in '
+                            'range [1, {}]'.format(self.MAX_PAGE_SIZE))
+
+        # Set up the SolveClient
+        # (kwargs overrides pre-set, which overrides global)
+        self._client = kwargs.get('client') or self._client or client
+
+    def _clone(self, filters=None, limit=None):
```

Comment:
Can we just use the parent class methods?

Reply:
As the parent class of `GlobalSearch` is `Query`, which has some additional attributes like required dataset id and optional parameters in constructors such as _genome_build, _annotator_params, _target_fields and others.

And the `_clone` method creates a new instance of that class, in our case a new instance of `GlobalSearch` (`self.__class__(params...)`). So because of that, we have to override that method and specify the correct arguments for the constructor for GlobalSearch class.

Reply:
Ah yea I see. Ok thats fine.